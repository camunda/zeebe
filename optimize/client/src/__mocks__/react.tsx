/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

// @ts-ignore
export * from 'react';
export {default} from 'react';

const outstandingEffects: (() => void)[] = [];
const outstandingCleanups: (() => void)[] = [];

export const useEffect = (fn: () => (() => void) | void) => outstandingEffects.push(fn);
export const runLastEffect = () => {
  if (outstandingEffects.length) {
    const cleanup = outstandingEffects.pop()?.();

    if (cleanup) {
      outstandingCleanups.push(cleanup);
    }
  }
};

export const runLastCleanup = () => {
  if (outstandingCleanups.length) {
    outstandingCleanups.pop()?.();
  }
};

export const runAllEffects = () => {
  const numberOfEffects = outstandingEffects.length;

  for (let i = 0; i < numberOfEffects; i++) {
    const cleanup = outstandingEffects.shift()?.();

    if (cleanup) {
      outstandingCleanups.push(cleanup);
    }
  }
};

beforeEach(() => {
  outstandingEffects.length = 0;
});

export const runAllCleanups = () => {
  const numberOfEffects = outstandingCleanups.length;

  for (let i = 0; i < numberOfEffects; i++) {
    outstandingCleanups.shift()?.();
  }
};
