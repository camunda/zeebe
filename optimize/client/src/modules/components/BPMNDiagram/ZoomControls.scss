@use '@carbon/styles/scss/spacing';

.ZoomControls {
  display: flex;
  flex-direction: column;
  position: absolute;
  right: spacing.$spacing-05;
  bottom: spacing.$spacing-05;
  z-index: var(--z-above);

  > span {
    background-color: var(--cds-background);

    &:first-of-type {
      margin-bottom: spacing.$spacing-02;
    }

    &:last-of-type button {
      border-top: none;
    }
  }
}
