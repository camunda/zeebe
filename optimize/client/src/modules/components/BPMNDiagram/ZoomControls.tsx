/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {IconButton} from '@carbon/react';
import {Add, Subtract, CenterCircle} from '@carbon/icons-react';

import {t} from 'translation';

import './ZoomControls.scss';
import {useMemo} from 'react';

const ZOOM_STEP_SIZE = 0.1;

interface ZoomControlsProps {
  fit: () => void;
  zoom: (stepSize: number) => void;
}

export default function ZoomControls({fit, zoom}: ZoomControlsProps) {
  const buttons = useMemo(
    () => [
      {
        onClick: fit,
        className: 'reset',
        icon: CenterCircle,
        description: t('common.zoomControls.resetZoom').toString(),
        key: 'reset',
      },
      {
        onClick: () => zoom(ZOOM_STEP_SIZE),
        className: 'zoomIn',
        icon: Add,
        description: t('common.zoomControls.zoomIn').toString(),
        key: 'zoomIn',
      },
      {
        onClick: () => zoom(-ZOOM_STEP_SIZE),
        className: 'zoomOut',
        icon: Subtract,
        description: t('common.zoomControls.zoomOut').toString(),
        key: 'zoomOut',
      },
    ],
    [fit, zoom]
  );

  return (
    <div className="ZoomControls">
      {buttons.map(({onClick, className, icon: Icon, description, key}) => (
        <IconButton
          key={key}
          size="sm"
          kind="tertiary"
          className={className}
          onClick={onClick}
          label={description}
          align="left"
        >
          <Icon />
        </IconButton>
      ))}
    </div>
  );
}
