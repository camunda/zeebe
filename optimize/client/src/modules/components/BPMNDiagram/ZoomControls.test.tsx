/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {shallow} from 'enzyme';
import {IconButton} from '@carbon/react';

import ZoomControls from './ZoomControls';

it('should match snapshot', () => {
  const node = shallow(<ZoomControls zoom={jest.fn()} fit={jest.fn()} />);

  expect(node).toMatchSnapshot();
});

it('should invoke zoom function on zoom button click', async () => {
  const spy = jest.fn();
  const node = shallow(<ZoomControls zoom={spy} fit={jest.fn()} />);

  node.find(IconButton).at(1).simulate('click');

  expect(spy).toHaveBeenCalled();
});

it('should invoke fit function when clicking reset button', async () => {
  const spy = jest.fn();
  const node = shallow(<ZoomControls zoom={jest.fn()} fit={spy} />);

  node.find(IconButton).first().simulate('click');

  expect(spy).toHaveBeenCalled();
});
