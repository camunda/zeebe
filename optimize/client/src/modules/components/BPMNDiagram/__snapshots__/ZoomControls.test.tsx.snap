// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  className="ZoomControls"
>
  <ForwardRef(IconButton)
    align="left"
    className="reset"
    key="reset"
    kind="tertiary"
    label="Reset diagram zoom"
    onClick={[MockFunction]}
    size="sm"
  >
    <ForwardRef(CenterCircle) />
  </ForwardRef(IconButton)>
  <ForwardRef(IconButton)
    align="left"
    className="zoomIn"
    key="zoomIn"
    kind="tertiary"
    label="Zoom in diagram"
    onClick={[Function]}
    size="sm"
  >
    <ForwardRef(Add) />
  </ForwardRef(IconButton)>
  <ForwardRef(IconButton)
    align="left"
    className="zoomOut"
    key="zoomOut"
    kind="tertiary"
    label="Zoom out diagram"
    onClick={[Function]}
    size="sm"
  >
    <ForwardRef(Subtract) />
  </ForwardRef(IconButton)>
</div>
`;
